<a href="https://colab.research.google.com/github/HoangHungLN/MachineLearning_Assigment/blob/main/Assignment1_CEML2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [34]:
##Xử lý dữ liệu
  #Thêm thư viện
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display
plt.style.use("seaborn-v0_8")
sns.set_palette("Set2")

dataset_url = "https://raw.githubusercontent.com/HoangHungLN/MachineLearning_Assigment/refs/heads/main/mobiles_uncleaned.csv"
raw_data = pd.read_csv(dataset_url)

  #in ra vài dòng của dữ liệu
raw_data.head(5)

name   price  spec_score  user_rating  \
0  Samsung Galaxy S24 FE 5G   34999        84.0         4.00   
1  Samsung Galaxy S24 Ultra   84949        92.0         4.35   
2   Samsung Galaxy S25 Edge  109999        89.0         4.45   
3     Samsung Galaxy M36 5G   19990        82.0         4.55   
4     Samsung Galaxy M06 5G    7999        72.0         4.00   

                                    connectivity  \
0  Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
1  Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
2  Dual Sim, 3G, 4G, 5G, VoLTE, Vo5G, Wi-Fi, NFC   
3        Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi, NFC   
4             Dual Sim, 3G, 4G, 5G, VoLTE, Wi-Fi   

                                                cpu  \
0       Exynos 2400e, Deca Core, 3.11 GHz Processor   
1   Snapdragon 8 Gen3, Octa Core, 3.3 GHz Processor   
2  Snapdragon 8Elite, Octa Core, 4.47 GHz Processor   
3         Exynos 1380, Octa Core, 2.4 GHz Processor   
4      Dimensity 6300, Octa Core, 2.4 GHz Processor   

                         ram                                  battery  \
0   8 GB RAM, 128 GB inbuilt  4700 mAh Battery with 25W Fast Charging   
1  12 GB RAM, 256 GB inbuilt  5000 mAh Battery with 45W Fast Charging   
2  12 GB RAM, 256 GB inbuilt  3900 mAh Battery with 25W Fast Charging   
3   8 GB RAM, 128 GB inbuilt  6500 mAh Battery with 25W Fast Charging   
4   4 GB RAM, 128 GB inbuilt  5000 mAh Battery with 25W Fast Charging   

                                             display  \
0  6.7 inches, 1080 x 2340 px, 120 Hz Display wit...   
1  6.8 inches, 1440 x 3120 px, 120 Hz Display wit...   
2  6.7 inches, 1440 x 3120 px, 120 Hz Display wit...   
3  6.74 inches, 1080 x 2340 px, 120 Hz Display wi...   
4  6.74 inches, 720 x 1600 px, 90 Hz Display with...   

                                              camera  \
0  50 MP + 12 MP + 8 MP Triple Rear & 10 MP Front...   
1              200 MP Quad Rear & 12 MP Front Camera   
2      200 MP + 12 MP Dual Rear & 12 MP Front Camera   
3  50 MP + 12 MP + 5 MP Triple Rear & 16 MP Front...   
4         50 MP + 2 MP Dual Rear & 8 MP Front Camera   

                          expandable           os     fm_radio other  
0          Memory Card Not Supported  Android v14          NaN   NaN  
1          Memory Card Not Supported  Android v14          NaN   NaN  
2                                NaN  Android v15  No FM Radio   NaN  
3    Memory Card (Hybrid), upto 1 TB  Android v15          NaN   NaN  
4  Memory Card (Hybrid), upto 1.5 TB  Android v15          NaN   NaN

In [35]:
  #kiểm tra số lượng mẫu và thuộc tính gốc của dữ liệu
  print ("Số lượng mẫu có trong dataset là:", raw_data.shape[0])
  print ("Số lượng thuộc tính có trong dataset là:", raw_data.shape[1])


Số lượng mẫu có trong dataset là: 11786
Số lượng thuộc tính có trong dataset là: 14


Trong bộ dữ liệu này, các thuộc tính bị thiếu được kí hiệu là NULL, sau đây nhóm em sẽ tiến hành chuyển dữ liệu thiếu thành NaN để dễ thống kê

In [36]:
raw_data.replace("NULL", np.nan, inplace=True)
missing_count = raw_data.isna().sum()
print(missing_count)

name                0
price               0
spec_score          8
user_rating        20
connectivity       11
cpu              2157
ram              1261
battery           138
display            16
camera             20
expandable       1770
os               2963
fm_radio        10229
other           11785
dtype: int64


Để trực quan hóa thì nhóm sẽ tạo dữ liệu dạng bảng để thống kê các giá trị bị thiếu
  

In [44]:
missing_rate = (missing_count / len(raw_data) * 100).round(2)
missing_data = pd.DataFrame({'Missing Count': missing_count, 'Missing Rate': missing_rate}).sort_values(by='Missing Rate', ascending=False)
missing_data.index.name = 'Feature'
#in ra bảng giá trị missing value
missing_data

Missing Count  Missing Rate
Feature                                  
other                 11785         99.99
fm_radio              10229         86.79
os                     2963         25.14
cpu                    2157         18.30
expandable             1770         15.02
ram                    1261         10.70
battery                 138          1.17
user_rating              20          0.17
camera                   20          0.17
display                  16          0.14
connectivity             11          0.09
spec_score                8          0.07
name                      0          0.00
price                     0          0.00

In [45]:
print ("\n-----------------------------------------------------------------------------------Describe (numeric)--------------------------------------------------------------------------------------------")
mum_summary = raw_data.describe().round(2).T.reset_index()
mum_summary = mum_summary.rename(columns={'index': 'Feature'}).set_index('Feature')
mum_summary



-----------------------------------------------------------------------------------Describe (numeric)--------------------------------------------------------------------------------------------


count      mean       std    min      25%      50%      75%  \
Feature                                                                      
price        11786.0  17187.61  23707.84  178.0  3000.00  9490.00  20999.0   
spec_score   11778.0     51.94     26.66    3.0    24.00    58.00     76.0   
user_rating  11766.0      4.18      0.48    0.5     4.05     4.25      4.5   

                  max  
Feature                
price        480000.0  
spec_score       98.0  
user_rating       5.0

Khi chạy lệnh nêu các đặc trưng của các biến định lượng thì ra thấy được rằng


1.   Về thuộc tính giá (price)

  *   Giá trị trung bình của các điện thoại là 117187,61

  *   Độ lệch chuẩn của thuộc tính khá lớn (23707,84) cho thấy dữ liệu phân tán mạnh, không tập trung xung quanh giá trị trung bình
  
  * Nhiều khả năng tồn tại giá trị ngoại lai với mức giá quá cao hoặc quá thấp,dẫn đến phân phối bị lệch và kéo giá trị trung bình lên khá cao so với giá trị trung vị 9490


2.   Về thuộc tính điểm cấu hình (spec_score)

  * Giá trị trung bình của điểm cấu hình là 51,94 điểm, trung vị là 58 điểm
  * Độ lệch chuẩn là 26,66 được xem là khá lớn so với thang điểm 100, điều này có thấy dữ liệu trải rộng
  *Nhìn chung, vì giá trị trung bình nhỏ hơn trung vị nên phân phối có xu hướng lệch trái, nhưng vẫn cần sử dụng đồ thij để trực quan hơn

3. Về thuộc tính đánh giá của khách hàng (user_rating)
* Giá trị trung bình của đánh giá người dùng là 4.18 trên thang điểm 5, với trung vị là 4,25
* Độ lệch chuẩn nhỏ (0.48) cho thấy tập dữ liệu tập trung xung quanh giá trị trung bình
* Phân phối nghiên về các giá trị cao cho thấy điện thoại trong tập dữ liệu nhận được nhiều đánh giá tích cực từ người dùng, ít khi bị điểm thấp


In [46]:
#Thêm thư viện để hiển thị đẹp trong Colab

cat_summary = raw_data.describe(include='object').T.reset_index()  # transpose để cột thành hàng
cat_summary = cat_summary.rename(columns={'index': 'Feature'}).set_index('Feature') #đổi tên cột index thành Feature
print ("\n-----------------------------------------------------------------------------------Describe (categorical)--------------------------------------------------------------------------------------------")
cat_summary


-----------------------------------------------------------------------------------Describe (categorical)--------------------------------------------------------------------------------------------


count unique                                top  freq
Feature                                                            
name          11786  11468     Realme 9i 5G (6GB RAM + 128GB)     3
connectivity  11775     68     Dual Sim, 3G, 4G, VoLTE, Wi-Fi  2554
cpu            9629   1461       Quad Core, 1.3 GHz Processor   399
ram           10525    443           8 GB RAM, 128 GB inbuilt  1143
battery       11648    955                   1000 mAh Battery   695
display       11770   1751   2.4 inches, 240 x 320 px Display   681
camera        11766   1142                 0.3 MP Rear Camera  1227
expandable    10016     31  Memory Card Supported, upto 32 GB  2261
os             8823    184                        Android v11   818
fm_radio       1557      1                        No FM Radio  1557
other             1      1                              Wi-Fi     1

Khi chạy thống kê mô tả của các biến định tính ta thấy rằng:
* _name_ : có 11,468 giá trị duy nhất trên 11,786 dòng (gần như mỗi dòng là một model khác nhau). Điều này làm cột này không hữu ích để phân tích thống kê hoặc mô hình hóa (có thể bỏ hoặc chỉ dùng để tham khảo).

* _connectivity_ : gồm 68 loại, trong đó phổ biến nhất là “Dual Sim, 3G, 4G, VoLTE, Wi-Fi” (≈ 2,554 bản ghi). Đây là một thuộc tính có thể mã hóa và sử dụng.

* _cpu_ : có 1,461 loại, phổ biến nhất là “Quad Core, 1.3 GHz Processor” (~399 bản ghi). Tuy nhiên cột này có nhiều missing values (chỉ 9,629/11,786 bản ghi).

* _ram_ : gồm 443 loại, phổ biến nhất “8 GB RAM, 128 GB inbuilt” (~1,143 bản ghi). Thuộc tính này khá chi tiết nhưng còn phân mảnh nhiều.

* _battery_ : gồm 955 loại, phổ biến nhất “1000 mAh Battery” (~695 bản ghi). Có nhiều kiểu pin khác nhau, dữ liệu phân tán.

* _display_ : gồm 1,751 loại, phổ biến nhất “2.4 inches, 240x320px Display” (~681 bản ghi). Đa dạng, có thể cần chuẩn hóa (inch + độ phân giải tách riêng).

* _camera_ : gồm 1,142 loại, phổ biến nhất “0.3 MP Rear Camera” (~1,227 bản ghi). Dữ liệu có xu hướng cụ thể nhưng nhiều dạng trình bày khác nhau.

* _expandable_ : gồm 31 loại, phổ biến nhất “Memory Card Supported, upto 32GB” (~2,261 bản ghi). Đây là một biến categorical khá gọn.

* _os_ : gồm 184 loại, phổ biến nhất “Android v11” (~818 bản ghi). Có nhiều phiên bản Android khác nhau, cần chuẩn hóa nhóm.

* _fm_radio_ : dữ liệu khá thiếu, chỉ 1,557 bản ghi có thông tin (toàn bộ là “No FM Radio”). Giá trị phân loại này không đa dạng.

* _other_ : toàn bộ chỉ có một giá trị “Wi-Fi” → không có giá trị phân biệt, có thể bỏ khỏi mô hình

Qua đây ta có nhận xét rằng: Các thuộc tính như tên sản phẩm (name), fm radio và other có thể cân nhắc để loại bỏ ra khỏi mô hình. Ngoài ra cần chuẩn hóa các thuộc tính khác để có thể đem lại kết quả tốt nhất khi huấn luyện mô hình